File: 2. Clean Architecture Objective-C/Router.xctemplate/___FILEBASENAME___Router.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  Copyright (c) ___YEAR___ ___ORGANIZATIONNAME___. All rights reserved.
//

#import "___VARIABLE_sceneName:identifier___Router.h"
#import "___VARIABLE_sceneName:identifier___BoundaryModels.h"
#import "___VARIABLE_sceneName:identifier___ViewController.h"

@implementation ___VARIABLE_sceneName:identifier___Router
#pragma mark Manual navigation
- (void)navigateToSomewhere {
    // NOTE: Teach the router how to navigate to another scene. Some examples follow:
    
    // 1. Trigger a storyboard segue
    // [self.viewController performSegueWithIdentifier:@"ShowSomewhereScene" sender:nil];
    
    // 2. Present another view controller programmatically
    // [self.viewController presentViewController:someWhereViewController animated:YES completion:nil];
    
    // 3. Ask the navigation controller to push another view controller onto the stack
    // [[self.viewController navigationController] pushViewController:someWhereViewController animated:YES];
    
    // 4. Present a view controller from a different storyboard
    // UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"OtherThanMain" bundle:nil];
    // UIViewController *someWhereViewController = [storyboard instantiateInitialViewController];
    // [[someWhereViewController navigationController] pushViewController:someWhereViewController animated:YES];
}

#pragma mark Scene communication
- (void)passDataBetweenScenesOfSegue:(nonnull UIStoryboardSegue *)segue {
    // NOTE: Teach the router which scenes it can communicate with
    if ([segue.identifier isEqualToString:@"ShowSomewhereScene"]) {
        [self sendDataToSomewhereSceneOfSegue:segue];
    }
}

- (void)sendDataToSomewhereSceneOfSegue:(nonnull UIStoryboardSegue *)segue {
    // NOTE: Teach the router how to pass data to the next scene
    // SomeWhereViewController *someWhereViewController = (SomeWhereViewController *)segue.destinationViewController;
    // someWhereviewController.objectId = viewController.objectId;
}

@end

File: 1. Clean Architecture Swift/Router.xctemplate/___FILEBASENAME___Router.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  Copyright (c) ___YEAR___ ___ORGANIZATIONNAME___. All rights reserved.
//

import UIKit

/**
    Class to route to new view controller when it's needed to display something that current view controller is not responsible for.
 */
class ___FILEBASENAMEASIDENTIFIER___Router {
    weak var viewController: ___FILEBASENAMEASIDENTIFIER___ViewController?
    
    // MARK: Manual navigation
    func navigateToSomewhere() {
        // NOTE: Teach the router how to navigate to another scene. Some examples follow:
        
        // 1. Trigger a storyboard segue
        // viewController?.performSegue(withIdentifier: "ShowSomewhereScene", sender: nil)
        
        // 2. Present another view controller programmatically
        // viewController?.present(someWhereViewController, animated: true, completion: nil)
        
        // 3. Ask the navigation controller to push another view controller onto the stack
        // viewController?.navigationController?.pushViewController(someWhereViewController, animated: true)
        
        // 4. Present a view controller from a different storyboard
        // let storyboard = UIStoryboard(name: "OtherThanMain", bundle: nil)
        // let someWhereViewController = storyboard.instantiateInitialViewController() as! SomeWhereViewController
        // viewController?.navigationController?.pushViewController(someWhereViewController, animated: true)
    }
    
    // MARK: Scene communication
    func passDataBetweenScenes(of segue: UIStoryboardSegue) {
        // NOTE: Teach the router which scenes it can communicate with
        if segue.identifier == "ShowSomewhereScene" {
            sendDataToSomewhereScene(segue)
        }
    }
    
    func sendDataToSomewhereScene(_ segue: UIStoryboardSegue) {
        // NOTE: Teach the router how to pass data to the next scene
        // let someWhereViewController = segue.destination as! SomeWhereViewController
        // someWhereviewController.objectId = viewController?.objectId
    }
}

// MARK: - Pass segue event to router for scene communication.
extension ___FILEBASENAMEASIDENTIFIER___ViewController {
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        router.passDataBetweenScenes(of: segue)
    }
}
